llvm/lib/Support/Compression.cpp

```diff
@@ -202,6 +207,71 @@ Error zstd::decompress(ArrayRef<uint8_t> Input,
   return E;
 }
 
+static void zstdContextFree(ZSTD_DCtx *Ctx) {
+  size_t RC = ::ZSTD_freeDCtx(Ctx);
+  assert(::ZSTD_isError(RC) == 0);
+}
+
+Error zstd::decompress(ArrayRef<uint8_t> Input,
+                       SmallVectorImpl<uint8_t> &Output) {
+  size_t Res = 0;
+  std::unique_ptr<::ZSTD_DCtx, void (*)(::ZSTD_DCtx *)> Ctx(::ZSTD_createDCtx(),
+                                                            zstdContextFree);
+  Res = ::ZSTD_initDStream(&*Ctx);
```

Comment:
Overall, where did you learn to use this API? It's super weird. A reference might help.

Reply:
Basically, by reading `zstd.h`.
Also, the API is somewhat similar to `gzip.h`, in both you do:
1. Initialize some context
2. Feed the context the input
3. Flush out anything that might be left in the internal state
4. free the state object